icontrol/authtoken.py

```diff
@@ -0,0 +1,137 @@
+                                 json=login_body,
+                                 verify=False,
+                                 auth=HTTPBasicAuth(self.username,
+                                                    self.password))
+        self.attempts += 1
+        if not response.ok or not hasattr(response, "json"):
+            error_message = '%s Unexpected Error: %s for uri: %s\nText: %r' %\
+                            (response.status_code,
+                             response.reason,
+                             response.url,
+                             response.text)
+            raise iControlUnexpectedHTTPError(error_message,
+                                              response=response)
+        respJson = response.json()
+
+        expiration_bigip, created_bigip = None, None
+        try:
+            self.token = respJson['token']['token']
+            expiration_bigip = int(respJson['token']['expirationMicros']) / \
+                1000000.0
+            created_bigip = int(respJson['token']['lastUpdateMicros']) / \
+                1000000.0
+        except KeyError:
+            error_message = \
+                '%s Unparseable Response: %s for uri: %s\nText: %r' %\
+                (response.status_code,
+                 response.reason,
+                 response.url,
+                 response.text)
+            raise iControlUnexpectedHTTPError(error_message,
+                                              response=response)
+
+        # Set our token expiration time.
+        # The expirationMicros field is when BIG-IP will expire the token
+        # relative to its local clock.  To avoid issues caused by incorrect
+        # clocks or network latency, we'll compute an expiration time that is
+        # referenced to our local clock, and expires slightly before the token
+        # should actually expire on BIG-IP
+
+        # Reference to our clock: compute for how long this token is valid as
+        # the difference between when it expires and when it was created,
+        # according to BIG-IP.
+        if expiration_bigip < created_bigip:
+            error_message = \
+                '%s Token already expired: %s for uri: %s\nText: %r' % \
+                (response.status_code,
+                 time.ctime(expiration_bigip),
+                 response.url,
+                 response.text)
+            raise iControlUnexpectedHTTPError(error_message,
+                                              response=response)
+        valid_duration = expiration_bigip - created_bigip
+
+        # Assign new expiration time that is 1 minute earlier than BIG-IP's
+        # expiration time, as long as that would still be at least a minute in
+        # the future.  This should account for clock skew between us and
+        # BIG-IP.  By default tokens last for 60 minutes so getting one every
+        # 59 minutes instead of 60 is harmless.
+        if valid_duration > 120.0:
```

Comment:
If the duration is less than 120 seconds, we will attempt to use it for its full duration.  Since we consider the start time to be the time that we first sent the request, that cannot be later than the time the server received the request and created the token.

So, if our clock and BIG-IP's clock are in perfect sync, no issues (and actually in the case of perfect sync, this "renew early" provision is unnecessary).  If our clock is faster than BIG-IP's, also no problem (we'll consider the token expired a little early).

If:
- The BIG-IP is configured to have token durations smaller than 120 seconds, more than 30x smaller than the default.
- Our clock runs slower: it loses X seconds per token duration, where X is the total time it takes us to prepare the request, send it over the network, and for BIG-IP to receive and parse that request and generate a token.
- Clients generate a request right at the end of this token duration.

then we will attempt to use an old token instead of getting a new one, and the request will fail.